File: docs/scripts_js/Contracts.js
import {Xlsx} from "@fslab/fsspreadsheet";
import fs from "fs";
import path from "path";

// Write

export function normalizePathSeparators (str) {
  const normalizedPath = path.normalize(str)
  return normalizedPath.replace(/\\/g, '/');
}


export async function fulfillWriteContract (basePath, contract) {
    function ensureDirectory (filePath) {
        let dirPath = path.dirname(filePath)
        if (!fs.existsSync(dirPath)){
            fs.mkdirSync(dirPath, { recursive: true });
        }
    }
    const p = path.join(basePath,contract.Path)
    if (contract.Operation = "CREATE") {
        if (contract.DTO == undefined) {
            ensureDirectory(p)
            fs.writeFileSync(p, "")
        } else if (contract.DTOType == "ISA_Assay" || contract.DTOType == "ISA_Assay" || contract.DTOType == "ISA_Investigation") {
            ensureDirectory(p)
            await Xlsx.toFile(p, contract.DTO)
        } else if (contract.DTOType == "PlainText") {
            ensureDirectory(p)
            fs.writeFileSync(p, contract.DTO)
        } else {
            console.log("Warning: The given contract is not a correct ARC write contract: ", contract)
        }
    }
}

// Read

export async function fulfillReadContract (basePath, contract) {
  async function fulfill() {
      const normalizedPath = normalizePathSeparators(path.join(basePath, contract.Path))
      switch (contract.DTOType) {
          case "ISA_Assay":
          case "ISA_Study":
          case "ISA_Investigation":
              let fswb = await Xlsx.fromXlsxFile(normalizedPath)
              return fswb
              break;
          case "PlainText":
              let content = fs.readFile(normalizedPath)
              return content
              break;
          default:
              console.log(`Handling of ${contract.DTOType} in a READ contract is not yet implemented`)
      }
  }
  if (contract.Operation == "READ") {
      return await fulfill()
  } else {
      console.error(`Error (fulfillReadContract): "${contract}" is not a READ contract`)
  }
}

File: docs/scripts_python/Contracts.py
import os
from arctrl import Contract, DTO_Spreadsheet, DTO_Text
from fsspreadsheet.xlsx import Xlsx
from pathlib import Path

#  Write

def fulfill_write_contract (basePath : str, contract : Contract) :
    def ensure_directory (filePath: Path) :
        if filePath.suffix or filePath.name.startswith(".") :
            filePath = filePath.parent    
        path_str = str(filePath)
        # Split the path into individual directories
        directories = path_str.split(os.path.sep)
        current_path = ""
        # Iterate through each directory in the path
        for directory in directories:
            # Append the current directory to the current path
            current_path = os.path.join(current_path, directory)
            # Check if the current path exists as a directory
            exists = os.path.exists(current_path)
            if not exists:
                os.makedirs(current_path)

        
    p = Path(basePath).joinpath(contract.Path)
    if contract.Operation == "CREATE" :
        if contract.DTO == None :
            ensure_directory(p)
            Path.write_text(p, "")
        elif contract.DTOType.name == "ISA_Assay" or contract.DTOType.name == "ISA_Study" or contract.DTOType.name == "ISA_Investigation" :
            ensure_directory(p)
            Xlsx.to_xlsx_file(p, contract.DTO.item)
        elif contract.DTOType.name == "PlainText" :
            ensure_directory(p)
            Path.write_text(p, contract.DTO.item)
        else :
            print("Warning: The given contract is not a correct ARC write contract: ", contract)      
    

#  Read

def fulfill_read_contract (basePath : str, contract : Contract) :
    if contract.Operation == "READ" :
        normalizedPath = os.path.normpath(Path(basePath).joinpath(contract.Path))
        if contract.DTOType.name == "ISA_Assay" or contract.DTOType.name == "ISA_Study" or contract.DTOType.name == "ISA_Investigation" :        
            fswb = Xlsx.from_xlsx_file(normalizedPath)
            contract.DTO = DTO_Spreadsheet(fswb)
        elif contract.DTOType.name == "PlainText" :
            content = Path.read_text(normalizedPath)
            contract.DTO = DTO_Text(content)
        else :
            print("Handling of ${contract.DTOType} in a READ contract is not yet implemented")

    else :
        print("Error (fulfillReadContract): ${contract} is not a READ contract")
